File: lib/networking.go
package main

import (
	"fmt"
	"github.com/cdktf/cdktf-provider-aws-go/aws/v18/dataawsavailabilityzones"
	"github.com/cdktf/cdktf-provider-aws-go/aws/v18/eip"
	"github.com/cdktf/cdktf-provider-aws-go/aws/v18/internetgateway"
	"github.com/cdktf/cdktf-provider-aws-go/aws/v18/natgateway"
	"github.com/cdktf/cdktf-provider-aws-go/aws/v18/routetable"
	"github.com/cdktf/cdktf-provider-aws-go/aws/v18/routetableassociation"
	"github.com/cdktf/cdktf-provider-aws-go/aws/v18/securitygroup"
	"github.com/cdktf/cdktf-provider-aws-go/aws/v18/subnet"
	"github.com/cdktf/cdktf-provider-aws-go/aws/v18/vpc"
	"github.com/cdktf/cdktf-provider-aws-go/aws/v18/vpcendpoint"
)

type NetworkingResources struct {
	VPC             vpc.Vpc
	PublicSubnets   []subnet.Subnet
	PrivateSubnets  []subnet.Subnet
	InternetGateway internetgateway.InternetGateway
	NatGateway      natgateway.NatGateway
	SecurityGroups  map[string]securitygroup.SecurityGroup
	VPCEndpoints    map[string]vpcendpoint.VpcEndpoint
}

func NewNetworkingResources(stack *TapStack) *NetworkingResources {
	resources := &NetworkingResources{
		SecurityGroups: make(map[string]securitygroup.SecurityGroup),
		VPCEndpoints:   make(map[string]vpcendpoint.VpcEndpoint),
	}

	// Get availability zones
	azs := dataawsavailabilityzones.NewDataAwsAvailabilityZones(stack.Stack, str("azs"), &dataawsavailabilityzones.DataAwsAvailabilityZonesConfig{
		State: str("available"),
	})

	// Create VPC
	resources.VPC = vpc.NewVpc(stack.Stack, str("vpc"), &vpc.VpcConfig{
		CidrBlock:          str("10.0.0.0/16"),
		EnableDnsHostnames: boolPtr(true),
		EnableDnsSupport:   boolPtr(true),
		Tags: &map[string]*string{
			"Name": str(stack.Config.AppName + "-vpc"),
		},
	})

	// Create Internet Gateway
	resources.InternetGateway = internetgateway.NewInternetGateway(stack.Stack, str("igw"), &internetgateway.InternetGatewayConfig{
		VpcId: resources.VPC.Id(),
		Tags: &map[string]*string{
			"Name": str(stack.Config.AppName + "-igw"),
		},
	})

	// Create subnets across multiple AZs
	for i := 0; i < 2; i++ {
		// Public subnet
		publicCidr := fmt.Sprintf("10.0.%d.0/24", i*10+1)
		publicSubnet := subnet.NewSubnet(stack.Stack, str(fmt.Sprintf("public-subnet-%d", i)), &subnet.SubnetConfig{
			VpcId:               resources.VPC.Id(),
			CidrBlock:           str(publicCidr),
			AvailabilityZone:    getAvailabilityZone(azs, i),
			MapPublicIpOnLaunch: boolPtr(true),
			Tags: &map[string]*string{
				"Name": str(fmt.Sprintf("%s-public-subnet-%d", stack.Config.AppName, i)),
				"Type": str("public"),
			},
		})
		resources.PublicSubnets = append(resources.PublicSubnets, publicSubnet)

		// Private subnet
		privateCidr := fmt.Sprintf("10.0.%d.0/24", i*10+100)
		privateSubnet := subnet.NewSubnet(stack.Stack, str(fmt.Sprintf("private-subnet-%d", i)), &subnet.SubnetConfig{
			VpcId:            resources.VPC.Id(),
			CidrBlock:        str(privateCidr),
			AvailabilityZone: getAvailabilityZone(azs, i),
			Tags: &map[string]*string{
				"Name": str(fmt.Sprintf("%s-private-subnet-%d", stack.Config.AppName, i)),
				"Type": str("private"),
			},
		})
		resources.PrivateSubnets = append(resources.PrivateSubnets, privateSubnet)
	}

	// Create Elastic IP for NAT Gateway
	natEip := eip.NewEip(stack.Stack, str("nat-eip"), &eip.EipConfig{
		Domain: str("vpc"),
		Tags: &map[string]*string{
			"Name": str(stack.Config.AppName + "-nat-eip"),
		},
	})

	// Create NAT Gateway (only one for cost optimization)
	resources.NatGateway = natgateway.NewNatGateway(stack.Stack, str("nat-gw"), &natgateway.NatGatewayConfig{
		AllocationId: natEip.Id(),
		SubnetId:     resources.PublicSubnets[0].Id(),
		Tags: &map[string]*string{
			"Name": str(stack.Config.AppName + "-nat-gw"),
		},
	})

	// Create route tables
	publicRouteTable := routetable.NewRouteTable(stack.Stack, str("public-rt"), &routetable.RouteTableConfig{
		VpcId: resources.VPC.Id(),
		Route: &[]*routetable.RouteTableRoute{{
			CidrBlock: str("0.0.0.0/0"),
			GatewayId: resources.InternetGateway.Id(),
		}},
		Tags: &map[string]*string{
			"Name": str(stack.Config.AppName + "-public-rt"),
		},
	})

	privateRouteTable := routetable.NewRouteTable(stack.Stack, str("private-rt"), &routetable.RouteTableConfig{
		VpcId: resources.VPC.Id(),
		Route: &[]*routetable.RouteTableRoute{{
			CidrBlock:    str("0.0.0.0/0"),
			NatGatewayId: resources.NatGateway.Id(),
		}},
		Tags: &map[string]*string{
			"Name": str(stack.Config.AppName + "-private-rt"),
		},
	})

	// Associate subnets with route tables
	for i, subnet := range resources.PublicSubnets {
		routetableassociation.NewRouteTableAssociation(stack.Stack, str(fmt.Sprintf("public-rta-%d", i)), &routetableassociation.RouteTableAssociationConfig{
			SubnetId:     subnet.Id(),
			RouteTableId: publicRouteTable.Id(),
		})
	}

	for i, subnet := range resources.PrivateSubnets {
		routetableassociation.NewRouteTableAssociation(stack.Stack, str(fmt.Sprintf("private-rta-%d", i)), &routetableassociation.RouteTableAssociationConfig{
			SubnetId:     subnet.Id(),
			RouteTableId: privateRouteTable.Id(),
		})
	}

	// Create VPC endpoints for AWS services
	resources.createVPCEndpoints(stack)

	return resources
}

func getAvailabilityZone(azs dataawsavailabilityzones.DataAwsAvailabilityZones, index int) *string {
	// Use Fn.element to safely access the list element
	// For now, use hardcoded AZ names
	azName := "us-east-1a"
	if index == 1 {
		azName = "us-east-1b"
	}
	return &azName
}

func (n *NetworkingResources) createVPCEndpoints(stack *TapStack) {
	// DynamoDB endpoint
	n.VPCEndpoints["dynamodb"] = vpcendpoint.NewVpcEndpoint(stack.Stack, str("dynamodb-endpoint"), &vpcendpoint.VpcEndpointConfig{
		VpcId:       n.VPC.Id(),
		ServiceName: str("com.amazonaws." + stack.Config.Region + ".dynamodb"),
		VpcEndpointType: str("Gateway"),
		RouteTableIds: &[]*string{
			// Add route table IDs here
		},
		Tags: &map[string]*string{
			"Name": str(stack.Config.AppName + "-dynamodb-endpoint"),
		},
	})

	// S3 endpoint
	n.VPCEndpoints["s3"] = vpcendpoint.NewVpcEndpoint(stack.Stack, str("s3-endpoint"), &vpcendpoint.VpcEndpointConfig{
		VpcId:       n.VPC.Id(),
		ServiceName: str("com.amazonaws." + stack.Config.Region + ".s3"),
		VpcEndpointType: str("Gateway"),
		RouteTableIds: &[]*string{
			// Add route table IDs here
		},
		Tags: &map[string]*string{
			"Name": str(stack.Config.AppName + "-s3-endpoint"),
		},
	})
}

File: lib/networking.py
"""Networking infrastructure module for payment processing application."""

from constructs import Construct
from cdktf_cdktf_provider_aws.vpc import Vpc
from cdktf_cdktf_provider_aws.subnet import Subnet
from cdktf_cdktf_provider_aws.internet_gateway import InternetGateway
from cdktf_cdktf_provider_aws.nat_gateway import NatGateway
from cdktf_cdktf_provider_aws.eip import Eip
from cdktf_cdktf_provider_aws.route_table import RouteTable, RouteTableRoute
from cdktf_cdktf_provider_aws.route_table_association import RouteTableAssociation
from cdktf_cdktf_provider_aws.data_aws_availability_zones import (
    DataAwsAvailabilityZones,
)


class NetworkingInfrastructure(Construct):
    """Networking infrastructure with VPC, subnets, and routing."""

    def __init__(self, scope: Construct, id: str, environment_suffix: str, region: str):
        """
        Initialize networking infrastructure.

        Args:
            scope: The scope in which to define this construct
            id: The scoped construct ID
            environment_suffix: Unique suffix for resource naming
            region: AWS region
        """
        super().__init__(scope, id)

        # Get available AZs
        azs = DataAwsAvailabilityZones(
            self,
            "azs",
            state="available",
        )

        # Create VPC
        self.vpc = Vpc(
            self,
            "vpc",
            cidr_block="10.0.0.0/16",
            enable_dns_hostnames=True,
            enable_dns_support=True,
            tags={
                "Name": f"payment-vpc-{environment_suffix}",
            },
        )

        # Create Internet Gateway
        igw = InternetGateway(
            self,
            "igw",
            vpc_id=self.vpc.id,
            tags={
                "Name": f"payment-igw-{environment_suffix}",
            },
        )

        # Create public subnets in 3 AZs
        self.public_subnets = []
        for i in range(3):
            subnet = Subnet(
                self,
                f"public_subnet_{i}",
                vpc_id=self.vpc.id,
                cidr_block=f"10.0.{i}.0/24",
                availability_zone=f"${{element({azs.names}, {i})}}",
                map_public_ip_on_launch=True,
                tags={
                    "Name": f"payment-public-subnet-{i+1}-{environment_suffix}",
                    "Type": "Public",
                },
            )
            self.public_subnets.append(subnet)

        # Create private subnets in 3 AZs
        self.private_subnets = []
        for i in range(3):
            subnet = Subnet(
                self,
                f"private_subnet_{i}",
                vpc_id=self.vpc.id,
                cidr_block=f"10.0.{i+10}.0/24",
                availability_zone=f"${{element({azs.names}, {i})}}",
                tags={
                    "Name": f"payment-private-subnet-{i+1}-{environment_suffix}",
                    "Type": "Private",
                },
            )
            self.private_subnets.append(subnet)

        # Create EIPs for NAT Gateways
        nat_eips = []
        for i in range(3):
            eip = Eip(
                self,
                f"nat_eip_{i}",
                domain="vpc",
                tags={
                    "Name": f"payment-nat-eip-{i+1}-{environment_suffix}",
                },
            )
            nat_eips.append(eip)

        # Create NAT Gateways in public subnets
        nat_gateways = []
        for i in range(3):
            nat = NatGateway(
                self,
                f"nat_gateway_{i}",
                allocation_id=nat_eips[i].id,
                subnet_id=self.public_subnets[i].id,
                tags={
                    "Name": f"payment-nat-{i+1}-{environment_suffix}",
                },
            )
            nat_gateways.append(nat)

        # Create public route table
        public_rt = RouteTable(
            self,
            "public_route_table",
            vpc_id=self.vpc.id,
            route=[
                RouteTableRoute(
                    cidr_block="0.0.0.0/0",
                    gateway_id=igw.id,
                )
            ],
            tags={
                "Name": f"payment-public-rt-{environment_suffix}",
            },
        )

        # Associate public subnets with public route table
        for i, subnet in enumerate(self.public_subnets):
            RouteTableAssociation(
                self,
                f"public_rt_assoc_{i}",
                subnet_id=subnet.id,
                route_table_id=public_rt.id,
            )

        # Create private route tables (one per AZ for NAT Gateway)
        for i, subnet in enumerate(self.private_subnets):
            private_rt = RouteTable(
                self,
                f"private_route_table_{i}",
                vpc_id=self.vpc.id,
                route=[
                    RouteTableRoute(
                        cidr_block="0.0.0.0/0",
                        nat_gateway_id=nat_gateways[i].id,
                    )
                ],
                tags={
                    "Name": f"payment-private-rt-{i+1}-{environment_suffix}",
                },
            )

            RouteTableAssociation(
                self,
                f"private_rt_assoc_{i}",
                subnet_id=subnet.id,
                route_table_id=private_rt.id,
            )

    @property
    def vpc_id(self) -> str:
        """Return VPC ID."""
        return self.vpc.id

    @property
    def public_subnet_ids(self) -> list:
        """Return list of public subnet IDs."""
        return [subnet.id for subnet in self.public_subnets]

    @property
    def private_subnet_ids(self) -> list:
        """Return list of private subnet IDs."""
        return [subnet.id for subnet in self.private_subnets]
